man/man.roff/db_stat.1
.ds TYPE C
.\"
.\" See the file LICENSE for redistribution information.
.\"
.\" Copyright (c) 1996, 1997
.\"	Sleepycat Software.  All rights reserved.
.\"
.\"	@(#)db_stat.so	8.10 (Sleepycat) 10/14/97
.\"
.\"
.\" See the file LICENSE for redistribution information.
.\"
.\" Copyright (c) 1997
.\"	Sleepycat Software.  All rights reserved.
.\"
.\"	@(#)macros.so	10.38 (Sleepycat) 1/18/98
.\"
.\" We don't want hyphenation for any HTML documents.
.ie '\*[HTML]'YES'\{\
.nh
\}
.el\{\
.ds Hy
.hy
..
.ds Nh
.nh
..
\}
.\" The alternative text macro
.\" This macro takes two arguments:
.\"	+ the text produced if this is a "C" manpage
.\"	+ the text produced if this is a "CXX" or "JAVA" manpage
.\"
.de Al
.ie '\*[TYPE]'C'\{\\$1
\}
.el\{\\$2
\}
..
.\" Scoped name macro.
.\" Produces a_b, a::b, a.b depending on language
.\" This macro takes two arguments:
.\"	+ the class or prefix (without underscore)
.\"	+ the name within the class or following the prefix
.de Sc
.ie '\*[TYPE]'C'\{\\$1_\\$2
\}
.el\{\
.ie '\*[TYPE]'CXX'\{\\$1::\\$2
\}
.el\{\\$1.\\$2
\}
\}
..
.\" Scoped name for Java.
.\" Produces Db.b, for Java, otherwise just b.  This macro is used for
.\" constants that must be scoped in Java, but are global otherwise.
.\" This macro takes two arguments:
.\"	+ the class
.\"	+ the name within the class or following the prefix
.de Sj
.ie '\*[TYPE]'JAVA'\{\
.TP 5
Db.\\$1\}
.el\{\
.TP 5
\\$2\}
..
.\" The general information text macro.
.de Gn
.ie '\*[TYPE]'C'\{The DB library is a family of groups of functions that provides a modular
programming interface to transactions and record-oriented file access.
The library includes support for transactions, locking, logging and file
page caching, as well as various indexed access methods.
Many of the functional groups (e.g., the file page caching functions)
are useful independent of the other DB functions,
although some functional groups are explicitly based on other functional
groups (e.g., transactions and logging).
\}
.el\{The DB library is a family of classes that provides a modular
programming interface to transactions and record-oriented file access.
The library includes support for transactions, locking, logging and file
page caching, as well as various indexed access methods.
Many of the classes (e.g., the file page caching class)
are useful independent of the other DB classes,
although some classes are explicitly based on other classes
(e.g., transactions and logging).
\}
For a general description of the DB package, see
.IR db_intro (3).
..
.\" The library error macro, the local error macro.
.\" These macros take one argument:
.\"	+ the function name.
.de Ee
The
.I \\$1
.ie '\*[TYPE]'C'\{function may fail and return
.I errno
\}
.el\{method may fail and throw a
.IR DbException (3)
.if '\*[TYPE]'CXX'\{
or return
.I errno
\}
\}
for any of the errors specified for the following DB and library functions:
..
.de Ec
In addition, the
.I \\$1
.ie '\*[TYPE]'C'\{function may fail and return
.I errno
\}
.el\{method may fail and throw a
.IR DbException (3)
.ie '\*[TYPE]'CXX'\{or return
.I errno
\}
.el\{encapsulating an
.I errno
\}
\}
for the following conditions:
..
.de Ea
[EAGAIN]
A lock was unavailable.
..
.de Eb
[EBUSY]
The shared memory region was in use and the force flag was not set.
..
.de Em
[EAGAIN]
The shared memory region was locked and (repeatedly) unavailable.
..
.de Ei
[EINVAL]
An invalid flag value or parameter was specified.
..
.de Es
[EACCES]
An attempt was made to modify a read-only database.
..
.de Et
The DB_THREAD flag was specified and spinlocks are not implemented for
this architecture.
..
.de Ep
[EPERM]
Database corruption was detected.
All subsequent database calls (other than
.ie '\*[TYPE]'C'\{\
.IR DB->close )
\}
.el\{\
.IR Db::close )
\}
will return EPERM.
..
.de Ek
.if '\*[TYPE]'CXX'\{\
Methods marked as returning
.I errno
will, by default, throw an exception that encapsulates the error information.
The default error behavior can be changed, see
.IR DbException (3).
\}
..
.\" The SEE ALSO text macro
.de Sa
.\" make the line long for nroff.
.if n .ll 72
.nh
.na
.IR db_archive (1),
.IR db_checkpoint (1),
.IR db_deadlock (1),
.IR db_dump (1),
.IR db_load (1),
.IR db_recover (1),
.IR db_stat (1),
.IR db_intro (3),
.ie '\*[TYPE]'C'\{\
.IR db_appinit (3),
.IR db_cursor (3),
.IR db_dbm (3),
.IR db_internal (3),
.IR db_lock (3),
.IR db_log (3),
.IR db_mpool (3),
.IR db_open (3),
.IR db_thread (3),
.IR db_txn (3)
\}
.el\{\
.IR db_internal (3),
.IR db_thread (3),
.IR Db (3),
.IR Dbc (3),
.IR DbEnv (3),
.IR DbException (3),
.IR DbInfo (3),
.IR DbLock (3),
.IR DbLockTab (3),
.IR DbLog (3),
.IR DbLsn (3),
.IR DbMpool (3),
.IR DbMpoolFile (3),
.IR Dbt (3),
.IR DbTxn (3),
.IR DbTxnMgr (3)
\}
.ad
.Hy
..
.\" The function header macro.
.\" This macro takes one argument:
.\"	+ the function name.
.de Fn
.in 2
.I \\$1
.in
..
.\" The XXX_open function text macro, for merged create/open calls.
.\" This macro takes two arguments:
.\"	+ the interface, e.g., "transaction region"
.\"	+ the prefix, e.g., "txn" (or the class name for C++, e.g., "DbTxn")
.de Co
.ie '\*[TYPE]'C'\{\
.Fn \\$2_open
The
.I \\$2_open
function copies a pointer, to the \\$1 identified by the
.B directory
.IR dir ,
into the memory location referenced by
.IR regionp .
.PP
If the
.I dbenv
argument to
.I \\$2_open
was initialized using
.IR db_appinit ,
.I dir
is interpreted as described by
.IR db_appinit (3).
\}
.el\{\
.Fn \\$2::open
The
.I \\$2::open
.ie '\*[TYPE]'CXX'\{\
method copies a pointer, to the \\$1 identified by the
.B directory
.IR dir ,
into the memory location referenced by
.IR regionp .
\}
.el\{\
method returns a \\$1 identified by the
.B directory
.IR dir .
\}
.PP
If the
.I dbenv
argument to
.I \\$2::open
was initialized using
.IR DbEnv::appinit ,
.I dir
is interpreted as described by
.IR DbEnv (3).
\}
.PP
Otherwise,
if
.I dir
is not NULL,
it is interpreted relative to the current working directory of the process.
If
.I dir
is NULL,
the following environment variables are checked in order:
``TMPDIR'', ``TEMP'', and ``TMP''.
If one of them is set,
\\$1 files are created relative to the directory it specifies.
If none of them are set, the first possible one of the following
directories is used:
.IR /var/tmp ,
.IR /usr/tmp ,
.IR /temp ,
.IR /tmp ,
.I C:/temp
and
.IR C:/tmp .
.PP
All files associated with the \\$1 are created in this directory.
This directory must already exist when
.ie '\*[TYPE]'C'\{
\\$1_open
\}
.el\{\
\\$2::open
\}
is called.
If the \\$1 already exists,
the process must have permission to read and write the existing files.
If the \\$1 does not already exist,
it is optionally created and initialized.
..
.\" The common close language macro, for discarding created regions
.\" This macro takes one argument:
.\"	+ the function prefix, e.g., txn (the class name for C++, e.g., DbTxn)
.de Cc
In addition, if the
.I dir
argument to
.ie '\*[TYPE]'C'\{\
.ds Va db_appinit
.ds Vo \\$1_open
.ds Vu \\$1_unlink
\}
.el\{\
.ds Va DbEnv::appinit
.ds Vo \\$1::open
.ds Vu \\$1::unlink
\}
.I \\*(Vo
was NULL
and
.I dbenv
was not initialized using
.IR \\*(Va ,
.if '\\$1'memp'\{\
or the DB_MPOOL_PRIVATE flag was set,
\}
all files created for this shared region will be removed,
as if
.I \\*(Vu
were called.
.rm Va
.rm Vo
.rm Vu
..
.\" The DB_ENV information macro.
.\" This macro takes two arguments:
.\"	+ the function called to open, e.g., "txn_open"
.\"	+ the function called to close, e.g., "txn_close"
.de En
.ie '\*[TYPE]'C'\{\
based on the
.I dbenv
argument to
.IR \\$1 ,
which is a pointer to a structure of type DB_ENV (typedef'd in <db.h>).
It is expected that applications will use a single DB_ENV structure as the
argument to all of the subsystems in the DB package.
In order to ensure compatibility with future releases of DB, all fields of
the DB_ENV structure that are not explicitly set should be initialized to 0
before the first time the structure is used.
Do this by declaring the structure external or static, or by calling the C
library routine
.IR bzero (3)
or
.IR memset (3).
.PP
The fields of the DB_ENV structure used by
.I \\$1
are described below.
.if '\*[TYPE]'CXX'\{\
As references to the DB_ENV structure may be maintained by
.IR \\$1 ,
it is necessary that the DB_ENV structure and memory it references be valid
until the
.I \\$2
function is called.
\}
.ie '\\$1'db_appinit'\{The
.I dbenv
argument may not be NULL.
If any of the fields of the
.I dbenv
are set to 0,
defaults appropriate for the system are used where possible.
\}
.el\{If
.I dbenv
is NULL
or any of its fields are set to 0,
defaults appropriate for the system are used where possible.
\}
.PP
The following fields in the DB_ENV structure may be initialized before calling
.IR \\$1 :
\}
.el\{\
based on which set methods have been used.
It is expected that applications will use a single DbEnv object as the
argument to all of the subsystems in the DB package.
The fields of the DbEnv object used by
.I \\$1
are described below.
As references to the DbEnv object may be maintained by
.IR \\$1 ,
it is necessary that the DbEnv object and memory it references be valid
until the object is destroyed.
.ie '\\$1'appinit'\{\
The
.I dbenv
argument may not be NULL.
If any of the fields of the
.I dbenv
are set to 0,
defaults appropriate for the system are used where possible.
\}
.el\{\
Any of the DbEnv fields that are not explicitly set will default to
appropriate values.
\}
.PP
The following fields in the DbEnv object may be initialized, using the
appropriate set method, before calling
.IR \\$1 :
\}
..
.\" The DB_ENV common fields macros.
.de Se
.ie '\*[TYPE]'JAVA'\{\
.TP 5
DbErrcall db_errcall;
.ns
.TP 5
String db_errpfx;
.ns
.TP 5
int db_verbose;
The error fields of the DbEnv behave as described for
.IR DbEnv (3).
\}
.el\{\
.ie '\*[TYPE]'CXX'\{\
.TP 5
void *(*db_errcall)(char *db_errpfx, char *buffer);
.ns
.TP 5
FILE *db_errfile;
.ns
.TP 5
const char *db_errpfx;
.ns
.TP 5
class ostream *db_error_stream;
.ns
.TP 5
int db_verbose;
The error fields of the DbEnv behave as described for
.IR DbEnv (3).
\}
.el\{\
void *(*db_errcall)(char *db_errpfx, char *buffer);
.ns
.TP 5
FILE *db_errfile;
.ns
.TP 5
const char *db_errpfx;
.ns
.TP 5
int db_verbose;
The error fields of the DB_ENV behave as described for
.IR db_appinit (3).
\}
\}
..
.\" The open flags.
.de Fm
The
.I flags
and
.I mode
arguments specify how files will be opened and/or created when they
don't already exist.
The flags value is specified by
.BR or 'ing
together one or more of the following values:
.Sj DB_CREATE
Create any underlying files, as necessary.
If the files do not already exist and the DB_CREATE flag is not specified,
the call will fail.
..
.\" DB_THREAD open flag macro.
.\" This macro takes two arguments:
.\"	+ the open function name
.\"	+ the object it returns.
.de Ft
.TP 5
.Sj DB_THREAD
Cause the \\$2 handle returned by the
.I \\$1
.Al function method
to be useable by multiple threads within a single address space,
i.e., to be ``free-threaded''.
.if '\*[TYPE]'JAVA'\{\
Threading is assumed in the Java API,
so no special flags are required,
and DB functions will always behave as if the DB_THREAD flag was specified.
\}
..
.\" The mode macro.
.\" This macro takes one argument:
.\"	+ the subsystem name.
.de Mo
All files created by the \\$1 are created with mode
.I mode
(as described in
.IR chmod (2))
and modified by the process' umask value at the time of creation (see
.IR umask (2)).
The group ownership of created files is based on the system and directory
defaults, and is not further specified by DB.
..
.\" The application exits macro.
.\" This macro takes one argument:
.\"	+ the application name.
.de Ex
The
.I \\$1
utility exits 0 on success, and >0 if an error occurs.
..
.\" The application -h section.
.\" This macro takes one argument:
.\"	+ the application name
.de Dh
DB_HOME
If the
.B \-h
option is not specified and the environment variable
.I DB_HOME
is set, it is used as the path of the database home, as described in
.IR db_appinit (3).
..
.\" The function DB_HOME ENVIRONMENT VARIABLES section.
.\" This macro takes one argument:
.\"	+ the open function name
.de Eh
DB_HOME
If the
.I dbenv
argument to
.I \\$1
was initialized using
.IR db_appinit ,
the environment variable DB_HOME may be used as the path of the database
home for the interpretation of the
.I dir
argument to
.IR \\$1 ,
as described in
.IR db_appinit (3).
.if \\n(.$>1 \{Specifically,
.I \\$1

perl.BerkeleyDB/Changes
Revision history for Perl extension BerkeleyDB.

0.01  23 October 1997
	* first alpha release as BerkDB.

0.02  30 October 1997
	* renamed module to BerkeleyDB	
	* fixed a few bugs & added more tests

0.03  5 May 1998
	* fixed db_get with DB_SET_RECNO
	* fixed c_get with DB_SET_RECNO and DB_GET_RECNO
	* implemented BerkeleyDB::Unknown
	* implemented BerkeleyDB::Recno, including push, pop etc
	  modified the txn support. 

0.04  19 May 1998
	* Define DEFSV & SAVE_DEFSV if not already defined. This allows
	  the module to be built with Perl 5.004_04.

0.05  9 November 1998
	* Added a note to README about how to build Berkeley DB 2.x
	  when using HP-UX.
	* Minor modifications to get the module to build with DB 2.5.x

0.06  19 December 1998
	* Minor modifications to get the module to build with DB 2.6.x
	* Added support for DB 2.6.x's  Concurrent Access Method, DB_INIT_CDB.

0.07  21st September 1999
	* Numerous small bug fixes.
	* Added support for sorting duplicate values DB_DUPSORT.
	* Added support for DB_GET_BOTH & DB_NEXT_DUP.
	* Added get_dup (from DB_File).
	* beefed up the documentation.
	* Forgot to add the DB_INIT_CDB in BerkeleyDB.pm in previous release.
	* Merged the DBM Filter code from DB_File into BerkeleyDB.
	* Fixed a nasty bug where a closed transaction was still used with
	  with dp_put, db_get etc.
	* Added logic to gracefully close everything whenever a fatal error
	  happens. Previously the plug was just pulled.
	* It is now a fatal error to explicitly close an environment if there
	  is still an open database; a database when there are open cursors or
	  an open transaction; and a cursor if there is an open transaction.
	  Using object destruction doesn't have this issue, as object
	  references will ensure everything gets closed in the correct order.
	* The BOOT code now checks that the version of db.h & libdb are the
	  same - this seems to be a common problem on Linux.
	* MLDBM support added.  
	* Support for the new join cursor added.
	* Builds with Berkeley DB 3.x
   	* Updated dbinfo for Berkeley DB 3.x file formats.
	* Deprecated the TxnMgr class. As with Berkeley DB version 3,
	  txn_begin etc are now accessed via the environment object.
	
0.08  28nd November 1999
	* More documentation updates
	* Changed reference to files in /tmp in examples.t
	* Fixed a typo in softCrash that caused problems when building
	  with a thread-enabled Perl.
	* BerkeleyDB::Error wasn't initialised properly.
	* ANSI-ified all the static C functions in BerkeleyDB.xs
	* Added support for the following DB 3.x features:
	    + The Queue database type
	    + db_remove
	    + subdatabases 
	    + db_stat for Hash & Queue

0.09  29th November 1999
        * the queue.t & subdb.t test harnesses were outputting a few
          spurious warnings. This has been fixed.

0.10  8th December 1999
	* The DESTROY method was missing for BerkeleyDB::Env. This resulted in
	  a memory leak. Fixed.
	* If opening an environment or database failed, there was a small 
	  memory leak. This has been fixed. 
	* A thread-enabled Perl it could core when a database was closed. 
	  Problem traced to the strdup function.

0.11  4th June 2000
	* When built with Berkeley Db 3.x there can ba a clash with the close
	  macro.
	* Typo in the definition of DB_WRITECURSOR
	* The flags parameter wasn't getting sent to db_cursor
	* Plugged small mempory leak in db_cursor (DESTROY wasn't freeing
	  memory)
	* Can be built with Berkeley DB 3.1
   


db_stat/db_stat.c
/*-
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 1996, 1997
 *	Sleepycat Software.  All rights reserved.
 */

#include "config.h"

#ifndef lint
static const char copyright[] =
"@(#) Copyright (c) 1997\n\
	Sleepycat Software Inc.  All rights reserved.\n";
static const char sccsid[] = "@(#)db_stat.c	8.30 (Sleepycat) 1/8/98";
#endif

#ifndef NO_SYSTEM_INCLUDES
#include <sys/types.h>

#include <ctype.h>
#include <errno.h>
#include <signal.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#endif

#include "db_int.h"
#include "clib_ext.h"

typedef enum { T_NOTSET, T_DB, T_LOG, T_MPOOL, T_TXN } test_t;

void	btree_stats __P((DB *));
DB_ENV *db_init __P((char *, test_t));
void	hash_stats __P((DB *));
int	main __P((int, char *[]));
void	log_stats __P((DB_ENV *));
void	mpool_stats __P((DB_ENV *));
void	onint __P((int));
void	prflags __P((u_int32_t, const FN *));
void	txn_stats __P((DB_ENV *));
int	txn_compare __P((const void *, const void *));
void	usage __P((void));

int	 interrupted;
const char
	*progname = "db_stat";				/* Program name. */

int
main(argc, argv)
	int argc;
	char *argv[];
{
	extern char *optarg;
	extern int optind;
	DB *dbp;
	DB_ENV *dbenv;
	test_t ttype;
	int ch;
	char *db, *home;

	ttype = T_NOTSET;
	db = home = NULL;
	while ((ch = getopt(argc, argv, "d:h:lmt")) != EOF)
		switch (ch) {
		case 'd':
			db = optarg;
			ttype = T_DB;
			break;
		case 'h':
			home = optarg;
			break;
		case 'l':
			ttype = T_LOG;
			break;
		case 'm':
			ttype = T_MPOOL;
			break;
		case 't':
			ttype = T_TXN;
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (argc != 0 || ttype == T_NOTSET)
		usage();

	dbenv = db_init(home, ttype);

	(void)signal(SIGINT, onint);

	switch (ttype) {
	case T_DB:
		if ((errno = db_open(db, DB_UNKNOWN,
		    DB_RDONLY, 0, dbenv, NULL, &dbp)) != 0)
			return (1);
		switch (dbp->type) {
		case DB_BTREE:
		case DB_RECNO:
			btree_stats(dbp);
			break;
		case DB_HASH:
			hash_stats(dbp);
			break;
		case DB_UNKNOWN:
			abort();		/* Impossible. */
			/* NOTREACHED */
		}
		(void)dbp->close(dbp, 0);
		break;
	case T_LOG:
		log_stats(dbenv);
		break;
	case T_MPOOL:
		mpool_stats(dbenv);
		break;
	case T_TXN:
		txn_stats(dbenv);
		break;
	case T_NOTSET:
		abort();			/* Impossible. */
		/* NOTREACHED */
	}

	(void)db_appexit(dbenv);

	if (interrupted) {
		(void)signal(SIGINT, SIG_DFL);
		(void)raise(SIGINT);
		/* NOTREACHED */
	}
	return (0);
}

/*
 * btree_stats --
 *	Display btree/recno statistics.
 */
void
btree_stats(dbp)
	DB *dbp;
{
	static const FN fn[] = {
		{ DB_DUP,	"DB_DUP" },
		{ DB_FIXEDLEN,	"DB_FIXEDLEN" },
		{ DB_RECNUM,	"DB_RECNUM" },
		{ DB_RENUMBER,	"DB_RENUMBER" },
		{ 0 }
	};
	DB_BTREE_STAT *sp;

	if (dbp->stat(dbp, &sp, NULL, 0))
		err(1, "dbp->stat");

#define	PCT(f, t)							\
    (t == 0 ? 0 :							\
    (((double)((t * sp->bt_pagesize) - f) / (t * sp->bt_pagesize)) * 100))

	printf("%#lx\tBtree magic number.\n", (u_long)sp->bt_magic);
	printf("%lu\tBtree version number.\n", (u_long)sp->bt_version);
	prflags(sp->bt_flags, fn);
	if (dbp->type == DB_BTREE) {
#ifdef NOT_IMPLEMENTED
		printf("%lu\tMaximum keys per-page.\n", (u_long)sp->bt_maxkey);
#endif
		printf("%lu\tMinimum keys per-page.\n", (u_long)sp->bt_minkey);
	}
	if (dbp->type == DB_RECNO) {
		printf("%lu\tFixed-length record size.\n",
		    (u_long)sp->bt_re_len);
		if (isprint(sp->bt_re_pad))
			printf("%c\tFixed-length record pad.\n",
			    (int)sp->bt_re_pad);
		else
			printf("0x%x\tFixed-length record pad.\n",
			    (int)sp->bt_re_pad);
	}
	printf("%lu\tUnderlying tree page size.\n", (u_long)sp->bt_pagesize);
	printf("%lu\tNumber of levels in the tree.\n", (u_long)sp->bt_levels);
	printf("%lu\tNumber of keys in the tree.\n", (u_long)sp->bt_nrecs);
	printf("%lu\tNumber of tree internal pages.\n", (u_long)sp->bt_int_pg);
	printf("%lu\tNumber of tree leaf pages.\n", (u_long)sp->bt_leaf_pg);
	printf("%lu\tNumber of tree duplicate pages.\n",
	    (u_long)sp->bt_dup_pg);
	printf("%lu\tNumber of tree overflow pages.\n",
	    (u_long)sp->bt_over_pg);
	printf("%lu\tNumber of pages on the free list.\n",
	    (u_long)sp->bt_free);
	printf("%lu\tNumber of pages freed for reuse.\n",
	    (u_long)sp->bt_freed);
	printf("%lu\tNumber of bytes free in tree internal pages (%.0f%% ff)\n",
	    (u_long)sp->bt_int_pgfree,
	    PCT(sp->bt_int_pgfree, sp->bt_int_pg));
	printf("%lu\tNumber of bytes free in tree leaf pages (%.0f%% ff).\n",
	    (u_long)sp->bt_leaf_pgfree,
	    PCT(sp->bt_leaf_pgfree, sp->bt_leaf_pg));
printf("%lu\tNumber of bytes free in tree duplicate pages (%.0f%% ff).\n",
	    (u_long)sp->bt_dup_pgfree,
	    PCT(sp->bt_dup_pgfree, sp->bt_dup_pg));
printf("%lu\tNumber of bytes free in tree overflow pages (%.0f%% ff).\n",
	    (u_long)sp->bt_over_pgfree,
	    PCT(sp->bt_over_pgfree, sp->bt_over_pg));
	printf("%lu\tNumber of bytes saved by prefix compression.\n",
	    (u_long)sp->bt_pfxsaved);
	printf("%lu\tTotal number of tree page splits.\n",
	    (u_long)sp->bt_split);
	printf("%lu\tNumber of root page splits.\n", (u_long)sp->bt_rootsplit);
	printf("%lu\tNumber of fast splits.\n", (u_long)sp->bt_fastsplit);
	printf("%lu\tNumber of hits in tree fast-insert code.\n",
	    (u_long)sp->bt_cache_hit);
	printf("%lu\tNumber of misses in tree fast-insert code.\n",
	    (u_long)sp->bt_cache_miss);
	printf("%lu\tNumber of keys added.\n", (u_long)sp->bt_added);
	printf("%lu\tNumber of keys deleted.\n", (u_long)sp->bt_deleted);
}

/*
 * hash_stats --
 *	Display hash statistics.
 */
void
hash_stats(dbp)
	DB *dbp;
{
	COMPQUIET(dbp, NULL);

	return;
}

/*
 * log_stats --
 *	Display log statistics.
 */
void
log_stats(dbenv)
	DB_ENV *dbenv;
{
	DB_LOG_STAT *sp;

	if (log_stat(dbenv->lg_info, &sp, NULL))
		err(1, NULL);

	printf("%#lx\tLog magic number.\n", (u_long)sp->st_magic);
	printf("%lu\tLog version number.\n", (u_long)sp->st_version);
	printf("%#o\tLog file mode.\n", sp->st_mode);
	if (sp->st_lg_max % MEGABYTE == 0)
		printf("%luMb\tLog file size.\n",
		    (u_long)sp->st_lg_max / MEGABYTE);
	else if (sp->st_lg_max % 1024 == 0)
		printf("%luKb\tLog file size.\n", (u_long)sp->st_lg_max / 1024);
	else
		printf("%lu\tLog file size.\n", (u_long)sp->st_lg_max);
	printf("%luMb\tLog bytes written (+%lu bytes).\n",
	    (u_long)sp->st_w_mbytes, (u_long)sp->st_w_bytes);
	printf("%luMb\tLog bytes written since last checkpoint (+%lu bytes).\n",
	    (u_long)sp->st_wc_mbytes, (u_long)sp->st_wc_bytes);
	printf("%lu\tTotal log file writes.\n", (u_long)sp->st_wcount);
	printf("%lu\tTotal log file flushes.\n", (u_long)sp->st_scount);
	printf("%lu\tCurrent log file number.\n", (u_long)sp->st_cur_file);
	printf("%lu\tCurrent log file offset.\n", (u_long)sp->st_cur_offset);
	printf("%lu\tThe number of region locks granted without waiting.\n",
	    (u_long)sp->st_region_nowait);
	printf("%lu\tThe number of region locks granted after waiting.\n",
	    (u_long)sp->st_region_wait);
}

/*
 * mpool_stats --
 *	Display mpool statistics.
 */
void
mpool_stats(dbenv)
	DB_ENV *dbenv;
{
	DB_MPOOL_FSTAT **fsp;
	DB_MPOOL_STAT *gsp;

	if (memp_stat(dbenv->mp_info, &gsp, &fsp, NULL))
		err(1, NULL);

	printf("%lu\tCache size (%luK).\n",
	    (u_long)gsp->st_cachesize, (u_long)gsp->st_cachesize / 1024);
	printf("%lu\tRequested pages found in the cache",
	    (u_long)gsp->st_cache_hit);
	if (gsp->st_cache_hit + gsp->st_cache_miss != 0)
		printf(" (%.0f%%)", ((double)gsp->st_cache_hit /
		    (gsp->st_cache_hit + gsp->st_cache_miss)) * 100);
	printf(".\n");
	printf("%lu\tRequested pages mapped into the process' address space.\n",
	    (u_long)gsp->st_map);
	printf("%lu\tRequested pages not found in the cache.\n",
	    (u_long)gsp->st_cache_miss);
	printf("%lu\tPages created in the cache.\n",
	    (u_long)gsp->st_page_create);
	printf("%lu\tPages read into the cache.\n", (u_long)gsp->st_page_in);
	printf("%lu\tPages written from the cache to the backing file.\n",
	    (u_long)gsp->st_page_out);
	printf("%lu\tClean pages forced from the cache.\n",
	    (u_long)gsp->st_ro_evict);
	printf("%lu\tDirty pages forced from the cache.\n",
	    (u_long)gsp->st_rw_evict);
	printf("%lu\tDirty buffers written by trickle-sync thread.\n",
	    (u_long)gsp->st_page_trickle);
	printf("%lu\tCurrent clean buffer count.\n",
	    (u_long)gsp->st_page_clean);
	printf("%lu\tCurrent dirty buffer count.\n",
	    (u_long)gsp->st_page_dirty);
	printf("%lu\tNumber of hash buckets used for page location.\n",
	    (u_long)gsp->st_hash_buckets);
	printf("%lu\tTotal number of times hash chains searched for a page.\n",
	    (u_long)gsp->st_hash_searches);
	printf("%lu\tThe longest hash chain searched for a page.\n",
	    (u_long)gsp->st_hash_longest);
	printf(
	    "%lu\tTotal number of hash buckets examined for page location.\n",
	    (u_long)gsp->st_hash_examined);
	printf("%lu\tThe number of region locks granted without waiting.\n",
	    (u_long)gsp->st_region_nowait);
	printf("%lu\tThe number of region locks granted after waiting.\n",
	    (u_long)gsp->st_region_wait);

	for (; fsp != NULL && *fsp != NULL; ++fsp) {
		printf("%s\n", DB_LINE);
		printf("%s\n", (*fsp)->file_name);
		printf("%lu\tPage size.\n", (u_long)(*fsp)->st_pagesize);
		printf("%lu\tRequested pages found in the cache",
		    (u_long)(*fsp)->st_cache_hit);
		if ((*fsp)->st_cache_hit + (*fsp)->st_cache_miss != 0)
			printf(" (%.0f%%)", ((double)(*fsp)->st_cache_hit /
			    ((*fsp)->st_cache_hit + (*fsp)->st_cache_miss)) *
			    100);
		printf(".\n");
	printf("%lu\tRequested pages mapped into the process' address space.\n",
		    (u_long)(*fsp)->st_map);
		printf("%lu\tRequested pages not found in the cache.\n",
		    (u_long)(*fsp)->st_cache_miss);
		printf("%lu\tPages created in the cache.\n",
		    (u_long)(*fsp)->st_page_create);
		printf("%lu\tPages read into the cache.\n",
		    (u_long)(*fsp)->st_page_in);
	printf("%lu\tPages written from the cache to the backing file.\n",
		    (u_long)(*fsp)->st_page_out);
	}
}

/*
 * txn_stats --
 *	Display transaction statistics.
 */
void
txn_stats(dbenv)
	DB_ENV *dbenv;
{
	DB_TXN_STAT *tstat;
	u_int32_t i;
	const char *p;

	if (txn_stat(dbenv->tx_info, &tstat, NULL))
		err(1, NULL);

	p = tstat->st_last_ckp.file == 0 ?
	    "No checkpoint LSN." : "File/offset for last checkpoint LSN.";
	printf("%lu/%lu\t%s\n", (u_long)tstat->st_last_ckp.file,
	    (u_long)tstat->st_last_ckp.offset, p);
	p = tstat->st_pending_ckp.file == 0 ?
	    "No pending checkpoint LSN." :
	    "File/offset for last pending checkpoint LSN.";
	printf("%lu/%lu\t%s\n",
	    (u_long)tstat->st_pending_ckp.file,
	    (u_long)tstat->st_pending_ckp.offset, p);
	if (tstat->st_time_ckp == 0)
		printf("0\tNo checkpoint timestamp.\n");
	else
		printf("%.24s\tCheckpoint timestamp.\n",
		    ctime(&tstat->st_time_ckp));
	printf("%lx\tLast transaction ID allocated.\n",
	    (u_long)tstat->st_last_txnid);
	printf("%lu\tMaximum number of active transactions.\n",
	    (u_long)tstat->st_maxtxns);
	printf("%lu\tNumber of transactions begun.\n",
	    (u_long)tstat->st_nbegins);
	printf("%lu\tNumber of transactions aborted.\n",
	    (u_long)tstat->st_naborts);
	printf("%lu\tNumber of transactions committed.\n",
	    (u_long)tstat->st_ncommits);
	printf("%lu\tActive transactions.\n", (u_long)tstat->st_nactive);
	qsort(tstat->st_txnarray,
	    tstat->st_nactive, sizeof(tstat->st_txnarray[0]), txn_compare);
	for (i = 0; i < tstat->st_nactive; ++i)
		printf("\tid: %lx; initial LSN file/offest %lu/%lu\n",
		    (u_long)tstat->st_txnarray[i].txnid,
		    (u_long)tstat->st_txnarray[i].lsn.file,
		    (u_long)tstat->st_txnarray[i].lsn.offset);
}

int
txn_compare(a1, b1)
	const void *a1, *b1;
{
	const DB_TXN_ACTIVE *a, *b;

	a = a1;
	b = b1;

	if (a->txnid > b->txnid)
		return (1);
	if (a->txnid < b->txnid)
		return (-1);
	return (0);
}

/*
 * prflags --
 *	Print out flag values.
 */
void
prflags(flags, fn)
	u_int32_t flags;
	FN const *fn;
{
	const FN *fnp;
	int found;
	const char *sep;

	sep = " ";
	printf("Flags:");
	for (found = 0, fnp = fn; fnp->mask != 0; ++fnp)
		if (fnp->mask & flags) {
			printf("%s%s", sep, fnp->name);
			sep = ", ";
			found = 1;
		}
	printf("\n");
}

/*
 * db_init --
 *	Initialize the environment.
 */
DB_ENV *
db_init(home, ttype)
	char *home;
	test_t ttype;
{
	DB_ENV *dbenv;
	int flags;

	if ((dbenv = (DB_ENV *)malloc(sizeof(DB_ENV))) == NULL) {
		errno = ENOMEM;
		err(1, NULL);
	}

	/*
	 * Try and use the shared regions when reporting statistics on the
	 * DB databases, so our information is as up-to-date as possible,
	 * even if the mpool cache hasn't been flushed.  If that fails, we
	 * turn off the DB_INIT_MPOOL flag and try again.
	 */
	flags = DB_USE_ENVIRON;
	switch (ttype) {
	case T_DB:
	case T_MPOOL:
		flags |= DB_INIT_MPOOL;
		break;
	case T_LOG:
		flags |= DB_INIT_LOG;
		break;
	case T_TXN:
		flags |= DB_INIT_TXN;
		break;
	case T_NOTSET:
		abort();
		/* NOTREACHED */
	}

	/*
	 * If it works, we're done.  Set the error output options so that
	 * future errors are correctly reported.
	 */
	memset(dbenv, 0, sizeof(*dbenv));
	if ((errno = db_appinit(home, NULL, dbenv, flags)) == 0) {
		dbenv->db_errfile = stderr;
		dbenv->db_errpfx = progname;
		return (dbenv);
	}

	/* Turn off the DB_INIT_MPOOL flag if it's a database. */
	if (ttype == T_DB)
		flags &= ~DB_INIT_MPOOL;

	/* Set the error output options -- this time we want a message. */
	memset(dbenv, 0, sizeof(*dbenv));
	dbenv->db_errfile = stderr;
	dbenv->db_errpfx = progname;

	/* Try again, and it's fatal if we fail. */
	if ((errno = db_appinit(home, NULL, dbenv, flags)) != 0)
		err(1, "db_appinit");

	return (dbenv);
}

/*
 * oninit --
 *	Interrupt signal handler.
 */
void
onint(signo)
	int signo;
{
	COMPQUIET(signo, 0);

	interrupted = 1;
}

void
usage()
{
	fprintf(stderr, "usage: db_stat [-mlt] [-d file] [-h home]\n");
	exit (1);
}
